lib/Dialect/Comb/CombFolds.cpp

```diff
@@ -1116,6 +1118,191 @@ static bool canonicalizeOrOfConcatsWithCstOperands(OrOp op, size_t concatIdx1,
+            argChecks[arg].emplace_back(Interval{i, lhsBound, rhsBound});
+            continue;
+          }
+          if (lhsPred == ICmpPredicate::ult && rhsPred == ICmpPredicate::ugt) {
+            argChecks[arg].emplace_back(Interval{i, rhsBound, lhsBound});
+            continue;
+          }
+        }
+      }
+    }
+
+    if (auto cmpOp = input.getDefiningOp<ICmpOp>()) {
+      APInt v;
+      if (matchPattern(cmpOp.rhs(), m_RConstant(v))) {
+        // Find equality tests: x == n
+        if (cmpOp.predicate() == ICmpPredicate::eq) {
+          if (v.isZero())
+            argChecks[cmpOp.lhs()].emplace_back(
+                Interval{i, std::nullopt, v + 1});
+          else
+            argChecks[cmpOp.lhs()].emplace_back(Interval{i, v - 1, v + 1});
+          continue;
+        }
+        // Find upper bound tests: x < n
+        if (cmpOp.predicate() == ICmpPredicate::ult) {
+          argChecks[cmpOp.lhs()].emplace_back(Interval{i, std::nullopt, v});
+          continue;
+        }
+      }
+    }
+
+    keptOperands[i] = true;
+  }
+
+  // For each value, try to compress the associated checks.
+  llvm::SmallVector<Value> newOperands;
+  for (auto &[arg, checks] : argChecks) {
+    // Order the checks by their lower bounds.
+    std::stable_sort(checks.begin(), checks.end(), [&](auto &lhs, auto &rhs) {
+      if (auto lowerLHS = lhs.lowerBound)
+        if (auto lowerRHS = rhs.lowerBound)
+          return lowerLHS->ult(*lowerRHS);
+      return false;
+    });
+
+    // Find sequences of overlapping checks and compress them.
+    for (auto *it = checks.begin(); it != checks.end();) {
+      auto *begin = it++;
+
+      auto lowerBound = begin->lowerBound;
+      APInt upperBound = begin->upperBound;
+      auto *endCheck = begin;
+      while (it != checks.end()) {
+        if (auto lowerBound = it->lowerBound) {
+          if (!lowerBound->ult(upperBound))
+            break;
+        }
+
+        APInt itBound = it->upperBound;
+        if (itBound.ugt(upperBound)) {
+          upperBound = itBound;
+          endCheck = it;
+        }
+        ++it;
+      }
+
+      // If there is no overlap or the range consists of only two
+      // consecutive numbers, do not create a new range.
+      size_t n = std::distance(begin, it);
+      if (n == 1 ||
+          (n == 2 && begin->isEqCheck() && std::next(begin)->isEqCheck())) {
+        while (begin != it) {
+          keptOperands[begin->index] = true;
+          ++begin;
+        }
+        continue;
+      }
+
+      // Fuse the locations of all the operands.
+      llvm::SmallVector<Location> locations;
+      for (auto *op = begin; op != it; ++op)
+        locations.push_back(inputs[op->index].getLoc());
+
+      LocationAttr loc = rewriter.getFusedLoc(locations);
+
+      // Build the check for the upper bound.
+      Value upperBoundCheck = rewriter.create<ICmpOp>(
+          loc, rewriter.getI1Type(), ICmpPredicate::ult, arg,
```

Comment:
A range with an equality check at the max value will produce a comparison at 1 + max, which isn't representable in the same size.  a ule would get around this.

Reply:
I'm still emitting ULT and UGT to avoid further trivial invocations to the canonicalizer, but I've added a missing guard for overflow.